src/gtest/test_univalue.cpp

```diff
@@ -0,0 +1,316 @@
+    EXPECT_EQ(arr[999].getValStr(), "");
+
+    arr.clear();
+    EXPECT_TRUE(arr.empty());
+    EXPECT_EQ(arr.size(), 0);
+}
+
+TEST(test_univalue, univalue_object)
+{
+    UniValue obj(UniValue::VOBJ);
+    string strKey, strVal;
+    UniValue v;
+
+    strKey = "age";
+    v.setInt(100);
+    EXPECT_TRUE(obj.pushKV(strKey, v));
+
+    strKey = "first";
+    strVal = "John";
+    EXPECT_TRUE(obj.pushKV(strKey, strVal));
+
+    strKey = "last";
+    const char *cVal = "Smith";
+    EXPECT_TRUE(obj.pushKV(strKey, cVal));
+
+    strKey = "distance";
+    EXPECT_TRUE(obj.pushKV(strKey, (int64_t) 25));
+
+    strKey = "time";
+    EXPECT_TRUE(obj.pushKV(strKey, (uint64_t) 3600));
+
+    strKey = "calories";
+    EXPECT_TRUE(obj.pushKV(strKey, (int) 12));
+
+    strKey = "temperature";
+    EXPECT_TRUE(obj.pushKV(strKey, (double) 90.012));
+
+    UniValue obj2(UniValue::VOBJ);
+    EXPECT_TRUE(obj2.pushKV("cat1", 9000));
+    EXPECT_TRUE(obj2.pushKV("cat2", 12345));
+
+    EXPECT_TRUE(obj.pushKVs(obj2));
+
+    EXPECT_FALSE(obj.empty());
+    EXPECT_EQ(obj.size(), 9);
+
+    EXPECT_EQ(obj["age"].getValStr(), "100");
+    EXPECT_EQ(obj["first"].getValStr(), "John");
+    EXPECT_EQ(obj["last"].getValStr(), "Smith");
+    EXPECT_EQ(obj["distance"].getValStr(), "25");
+    EXPECT_EQ(obj["time"].getValStr(), "3600");
+    EXPECT_EQ(obj["calories"].getValStr(), "12");
+    EXPECT_EQ(obj["temperature"].getValStr(), "90.012");
+    EXPECT_EQ(obj["cat1"].getValStr(), "9000");
+    EXPECT_EQ(obj["cat2"].getValStr(), "12345");
+
+    EXPECT_EQ(obj["nyuknyuknyuk"].getValStr(), "");
+
+    EXPECT_TRUE(obj.exists("age"));
+    EXPECT_TRUE(obj.exists("first"));
+    EXPECT_TRUE(obj.exists("last"));
+    EXPECT_TRUE(obj.exists("distance"));
+    EXPECT_TRUE(obj.exists("time"));
+    EXPECT_TRUE(obj.exists("calories"));
+    EXPECT_TRUE(obj.exists("temperature"));
+    EXPECT_TRUE(obj.exists("cat1"));
+    EXPECT_TRUE(obj.exists("cat2"));
+
+    EXPECT_TRUE(!obj.exists("nyuknyuknyuk"));
+
+    map<string, UniValue::VType> objTypes;
+    objTypes["age"] = UniValue::VNUM;
+    objTypes["first"] = UniValue::VSTR;
+    objTypes["last"] = UniValue::VSTR;
+    objTypes["distance"] = UniValue::VNUM;
+    objTypes["time"] = UniValue::VNUM;
+    objTypes["calories"] = UniValue::VNUM;
+    objTypes["temperature"] = UniValue::VNUM;
+    objTypes["cat1"] = UniValue::VNUM;
+    objTypes["cat2"] = UniValue::VNUM;
+    EXPECT_TRUE(obj.checkObject(objTypes));
+
+    objTypes["cat2"] = UniValue::VSTR;
+    EXPECT_TRUE(!obj.checkObject(objTypes));
+
+    obj.clear();
+    EXPECT_TRUE(obj.empty());
+    EXPECT_EQ(obj.size(), 0);
+}
+
+static const char *json1 =
+"[1.10000000,{\"key1\":\"str\\u0000\",\"key2\":800,\"key3\":{\"name\":\"martian http://test.com\"}}]";
+
+TEST(test_univalue, univalue_readwrite)
+{
+    UniValue v;
+    EXPECT_TRUE(v.read(json1));
+
+    string strJson1(json1);
+    EXPECT_TRUE(v.read(strJson1));
+
+    EXPECT_TRUE(v.isArray());
+    EXPECT_EQ(v.size(), 2);
```

Comment:
this usually gives a compilation warning.. if unsigned compared in EXPECT_EQ with signed.
add 'u' suffix in all places like this:
     EXPECT_EQ(v.size(), 2u)

Reply:
Fixed. Thanks